jf_agent/agent_logging.py

```diff
@@ -1,50 +1,91 @@
-from datetime import datetime
 from contextlib import contextmanager
 from functools import wraps
 import logging
 import os
-import traceback
+import sys
+import urllib3
 
 '''
 Guidance on logging/printing in the agent:
 
-stdout (e.g., print() output) is for the customer/operator; it should include
- - info on run progress: tqdm output, etc.
- - errors/warnings encountered
+stdout (e.g., print() output) is for the customer/operator; it should be used when
+we have customer sensitive data that they do not want to submit to jellyfish (i.e. passwords)
 
-logger output is for Jellyfish; it should include
+logger output is for Jellyfish and customers; it should include:
  - function entry/exit timings
  - loop iteration timings/liveness indicators
  - errors/warnings encountered
 
 logger output is submitted to Jellyfish, so should only contain non-sensitive data:
 e.g., function names, iteration counts
-
-Since we generally want errors/warnings to go to BOTH stdout and the logger, we
-should generally use log_and_print() instead of logger.whatever().
 '''
 
 LOG_FILE_NAME = 'jf_agent.log'
 
+# For styling in log files, I think it's best to always use new lines even when we use
+# the special character to ignore them. Leverage always_use_newlines for this
+def _emit_helper(_self: logging.Handler, record: logging.LogRecord, always_use_newlines=False):
+    special_code = '[!n]'
+
+    msg = _self.format(record)
+    if special_code in msg:
+        msg = msg.replace('[!n]', '')
+        if always_use_newlines:
+            msg += '\n'
+    else:
+        msg += '\n'
+
+    _self.stream.write(msg)
```

Comment:
Q: Worthwhile to add error handling here?

Reply:
Also a good idea